### DISCUSSION TITLE
What is the best restart order for mon and mgr when upgrading?

### DISCUSSION THREAD

**Discussion by microyahoo** [rook/rook#11860] [Q&A] [answered]
When we do a rook upgrade, ceph components will be restarted randomly(please forgive me if the restart order has been controlled in the code), such as mon, osd, mgr, mds, etc. In order to be as unaware of the business as possible, or to minimize the impact on the business, can the user control the restart order of certain components? So my question is, what is the best restart order for mon and mgr? Because the relationship between mons is leader and followers, leader transfer may cause IO blocking, and the relationship between mgrs is active and standbys. 

Regarding the restart of ceph components, there are many issues and discussions involved, as shown below:
- [Update the image version of operator, all daemons of the ceph will be restarted?](https://github.com/rook/rook/issues/11657)
    #11674
    #11760
- [mgr: upgrade image version of the rook operator, mgr will restart](https://github.com/rook/rook/issues/11799)
- [core: operator to skip reconcile of mgrs and mds](https://github.com/rook/rook/pull/11802)
- [Update the image version of operator, all daemons of the ceph will be restarted?](https://github.com/rook/rook/discussions/11639)
- [how to set "SkipReconcileLabelKey = "ceph.rook.io/do-not-reconcile"" to osd's deployment?](https://github.com/rook/rook/discussions/11689)
- [need to shrink ceph-mon from 3 down to 1](https://github.com/rook/rook/discussions/11476)
- [Switch rook-ceph host network online](https://github.com/rook/rook/issues/11748)
- [Switch to host networking on a running cluster](https://github.com/rook/rook/issues/11121#top)

**Comment by travisn** [comment, marked as answer]
The upgrade design doc can be found here for some background on the intended approach: https://github.com/rook/rook/blob/master/design/ceph/upgrade.md

The upgrade implementation has changed in a few ways since that design doc was written, but the overall goals are the same. To clarify some implementation and answer your questions, here are a few points. 

These daemons are updated sequentially by the main cluster controller:
1. The mons will first be updated in rolling fashion, one at a time. Each mon must be up and in quorum before continuing the upgrade of the next mon.
2. The mgrs will be updated in rolling fashion, one at a time. Each mgr must be online before continuing the upgrade of the next mgr.
3. The OSDs will be updated. The operator may choose to update multiple in parallel, as long as they are in the same failure domain.

Other daemons are updated in parallel by their controllers: rgw, mds, rbd mirror, csi driver, etc. 

There have been discussions about ordering in the past. It's accepted with ceph maintainers that this ordering is expected: mons, then mgr, then osds. If there are specific concerns about the ordering, it would be good to clarify with ceph maintainers.
